src/unistd.rs

```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
-/// store it in a provided buffer.  The buffer will be populated with bytes up
-/// to the length of the provided slice including a NUL terminating byte.  If
-/// the hostname is longer than the length provided, no error will be provided.
-/// The posix specification does not specify whether implementations will
-/// null-terminate in this case, but the nix implementation will ensure that the
-/// buffer is null terminated in this case.
+/// This function call attempts to get the host name for the running system
+/// and return it as a NUL-terminated string.
+///
+/// # Examples
 ///
 /// ```no_run
 /// use nix::unistd;
 ///
-/// let mut buf = [0u8; 64];
-/// let hostname_cstr = unistd::gethostname(&mut buf).expect("Failed getting hostname");
-/// let hostname = hostname_cstr.to_str().expect("Hostname wasn't valid UTF-8");
+/// let hostname_cstr = unistd::gethostname().expect("Failed getting hostname");
+/// let hostname = hostname_cstr.into_string().expect("Hostname wasn't valid UTF-8");
 /// println!("Hostname: {}", hostname);
 /// ```
-pub fn gethostname(buffer: &mut [u8]) -> Result<&CStr> {
-    let ptr = buffer.as_mut_ptr() as *mut c_char;
-    let len = buffer.len() as size_t;
+pub fn gethostname() -> Result<CString> {
+    // Hostname maximum length as defined by POSIX,
+    // http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html
+    const POSIX_HOST_NAME_MAX: usize = 255;
+    let buf = vec![0; POSIX_HOST_NAME_MAX + 1];
+    let ptr = unsafe { CString::from_vec_unchecked(buf).into_raw() };
+
+    let res = unsafe { libc::gethostname(ptr, POSIX_HOST_NAME_MAX + 1) };
+    Errno::result(res).map(|_| {
+        unsafe {
+            // Ensure returned string is always null-terminated
+            *ptr.offset(POSIX_HOST_NAME_MAX as isize) = 0;
+            CString::from_raw(ptr)
+        }
+    })
+}
 
-    let res = unsafe { libc::gethostname(ptr, len) };
+cfg_if!{
+    if #[cfg(any(target_os = "dragonfly", target_os = "freebsd",
+                 target_os = "ios", target_os = "macos"))] {
+        type namelen_t = c_int;
+    } else if #[cfg(not(target_os = "android"))] {
+        type namelen_t = size_t;
+    }
+}
+
+/// Set the NIS domain name (see
+/// [setdomainname(2)](http://man7.org/linux/man-pages/man2/setdomainname.2.html)).
+#[cfg(not(target_os = "android"))]
+pub fn setdomainname<S: AsRef<OsStr>>(name: S) -> Result<()> {
+    let ptr = name.as_ref().as_bytes().as_ptr() as *const c_char;
+    let len = name.as_ref().len() as namelen_t;
+    let res = unsafe { libc::setdomainname(ptr, len) };
+    Errno::result(res).map(drop)
+}
+
+/// Get the NIS domain name, returning a `CString` on success (see
+/// [getdomainname(2)](http://man7.org/linux/man-pages/man2/getdomainname.2.html)).
+///
+/// This function call attempts to get the NIS domain name for the running system
+/// and return it as a NUL-terminated string.
+///
+/// # Examples
+///
+/// ```no_run
```

Comment:
Can this not run in CI?

Reply:
Why use a `CString` here but an `OsStr` for `setdomainname`?  We should be consistent in our choice of C vs OS strings.

Reply:
@asomers This follows the split currently in master for the hostname, see `gethostname` vs `sethostname` types. Perhaps it can be left out of this PR? Otherwise, in which direction do you want to resolve this?